reference/reference.go

```diff
@@ -0,0 +1,283 @@
+// Package reference provides a general type to represent any way of referencing images within the registry.
+// Its main purpose is to abstract tags and digests (content-addressable hash).
+//
+// Grammar
+//
+// 	reference                       := repository [ ":" tag ] [ "@" digest ]
+//
+//	// repository.go
+//	repository			:= hostname ['/' component]+
+//	hostname 			:= hostcomponent [':' port-number]
+//	component			:= alpha-numeric [separator alpha-numeric]*
+//	hostcomponent                   := [hostpart '.']* hostpart
+// 	alpha-numeric			:= /[a-zA-Z0-9]+/
+//	separator			:= /[_-]/
```

Comment:
This seems inconsistent with the regexp. Should '.' be allowed in this part of the name, I do not believe it was allowed previously so that it did not interfere with hostname parsing.